### PR TITLE
Suggest an article on PrestaShop Circuit Breaker

### PR DESCRIPTION
**Pull request by mickaelandrieu**

### CODE REVIEW

File: news/_posts/2019-15-01-resilient-php-applications.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+date:   2019-01-10 10:00:00
+authors: [ mandrieu ]
+icon: icon-blaster
+tags:
+ - 1.7
+---
+
+Using an extensible e-commerce platform like PrestaShop is awesome for both developers and e-merchants. Relying on an open source project with the possibility to add themes and plugins available in the store allows every business to create a shop that fits their needs.
+
+But, no matter how well designed is your project and how many tests you have, at some point your website can become unavailable.
```
Comment (ttoine) [inline news/_posts/2019-15-01-resilient-php-applications.md]: this sentence must be changed a bit in order to be the first sentence, so it's displayed on the article intro in the blog feed.

File: news/_posts/2019-15-01-resilient-php-applications.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+This can be because of your database, or calls to third party services (payment methods API, PrestaShop REST API to retrieve modules information) are unavailable.
+
+Instead of thinking everything will be ok, we need - as developers - to manage the failures of our systems.
+
+We need to create resilient systems, to create websites able to recover quickly from network disasters.
+
+## The risks of depending to external services
```
Comment (ttoine) [inline news/_posts/2019-15-01-resilient-php-applications.md]:
```suggestion
## The risks of depending on external services
```

File: news/_posts/2019-15-01-resilient-php-applications.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+* you should be able to install PrestaShop even if the translations APIs are not reachable,
+at least in english;
+
+* you should be able to access your back office, no matter how many external APIs are not available;
+
+**The collapse of a third party service shouldn't collapse your shop.**
+
+Another issue with this kind of naive assumptions is that your system will try to contact again and again the third party service. This is slowing your system and make the user experience terrible when you could manage to provide an alternative response.
+
+> There is no good reason to provide terrible user experience! period.
```
Comment (ttoine) [inline news/_posts/2019-15-01-resilient-php-applications.md]:
```suggestion
> There is no good reason to provide such a terrible user experience! period.
```

File: news/_posts/2019-15-01-resilient-php-applications.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+This can be because of your database, or calls to third party services (payment methods API, PrestaShop REST API to retrieve modules information) are unavailable.
+
+Instead of thinking everything will be ok, we need - as developers - to manage the failures of our systems.
+
+We need to create resilient systems, to create websites able to recover quickly from network disasters.
+
+## The risks of depending to external services
+
+Assuming all external services will works fine is risky because you don't think
```
Comment (ttoine) [inline news/_posts/2019-15-01-resilient-php-applications.md]:
```suggestion
Assuming all external services will works fine is risky because you don't think about what should happens in case of unavailability:
```

File: news/_posts/2019-15-01-resilient-php-applications.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+about what should happens in case of unavailability:
```
Comment (ttoine) [inline news/_posts/2019-15-01-resilient-php-applications.md]:
```suggestion

```

File: news/_posts/2019-15-01-resilient-php-applications.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+* once we reach a defined numbers of failures (means the service is unreachable), the Circuit Breaker is "Open": an alternative response is provided and no calls to the service are done;
+* after a defined amount of time (called "reset timeout" sometimes), the Circuit Breaker try to access the service: if the service is available it will be used again, else the alternative response is provided instead;
+
+![https://martinfowler.com/bliki/images/circuitBreaker/sketch.png](https://martinfowler.com/bliki/images/circuitBreaker/sketch.png)
+
+> Schema from Martin Fowler's blog.
+
+This is also a big change of mindset for both the e-merchants and the developers, because for each process they need to think about what should be done in case of unavailability of a service.
+
+Like we already do in some old web browsers for the design of our websites, we should think about degraded experience when some third party services we use are not availables.
```
Comment (ttoine) [inline news/_posts/2019-15-01-resilient-php-applications.md]: who is "we"?
in general, please try to avoid to use "we" when it's not clear who is behind, and find a different way to write the sentence.
Example: "Like good developers do to handle old web browsers when they design a website, it's possible to think from the beginning about a "degraded experience" when some third party services are not available".

File: news/_posts/2019-15-01-resilient-php-applications.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+$circuitBreaker->call('https://api.domain.com', $fallbackResponse);
+```
+
+## What’s next?
+
+We will start to use it on the "so well-known" _gamification_ module soon,
+and the library will be available in PrestaShop in the next minor release (1.7.6). Step by step, we will make PrestaShop Core and native modules resilient to networks failures.
+
+As most of our work, we have released the circuit breaker under MIT license and on our main PrestaShop organization. You are very welcome to contribute the library and to fill issues if you have specific needs.
```
Comment (ttoine) [inline news/_posts/2019-15-01-resilient-php-applications.md]: Most of PrestaShop's developers work is actually released under OSL and AFL.

"The Circuit Breaker library is developed and distributed under the MIT license, like many other librairies and dependencies included in the PrestaShop project (example: Symfony). This way, it is possible for anyone to use it with any other project. Your feedback and your contribution are welcome: feel free to open pull requests and issues about your needs.

**Comment by ttoine** [comment]
The travis check is failing, could you please investigate why?

Also, this article needs a proofreading by a good English writer.

**Review by ttoine — CHANGES_REQUESTED** [review]
The article is a good introduction. Please check my comments and suggestions, and then review the English with a proofreader.

File: news/_posts/2019-15-01-resilient-php-applications.md (hunk 1 above)
Comment (mickaelandrieu) [reply news/_posts/2019-15-01-resilient-php-applications.md]: Agree, do you have a suggestion? I'm not really good to market my content ^^

File: news/_posts/2019-15-01-resilient-php-applications.md (hunk 6 above)
Comment (mickaelandrieu) [reply news/_posts/2019-15-01-resilient-php-applications.md]: I like your suggestion, but I don't want a fight between the "good" and the "bad" developers on my article 👍

WDYT of:


"While the web developers try to handle the old web browsers constraints when they design a website, it's possible to think from the beginning about a "degraded experience" when some third party services are not available with a circuit breaker."

File: news/_posts/2019-15-01-resilient-php-applications.md (hunk 1 above)
Comment (ttoine) [reply news/_posts/2019-15-01-resilient-php-applications.md]:
```suggestion
No matter how well designed is your project and how many tests you have, at some point your website can become unavailable.
```
let's kick the first sentence, this is not really useful here, for the concept and the library is not only about online stores and open source.

File: news/_posts/2019-21-01-resilient-php-applications.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+icon: icon-blaster
+tags:
+ - 1.7
+---
+
+Using an extensible e-commerce platform like PrestaShop is awesome for both developers and e-merchants. Relying on an open source project with the possibility to add themes and plugins available in the store allows every business to create a shop that fits their needs.
+
+But, no matter how well designed is your project and how many tests you have, at some point your website can become unavailable.
+
+This can be because of your database, or calls to third party services (payment methods API, PrestaShop REST API to retrieve modules information) are unavailable.
```
Comment (ttoine) [inline news/_posts/2019-21-01-resilient-php-applications.md]:
```suggestion
This can be because of your database, or calls to third party services are unavailable (e.g: payment methods API, PrestaShop REST API to retrieve modules information, synchronization with an ERP, automatisation of warehouse & logistic, ...)
```

File: news/_posts/2019-15-01-resilient-php-applications.md (hunk 6 above)
Comment (ttoine) [reply news/_posts/2019-15-01-resilient-php-applications.md]: "Like web developers do to handle old web browers when they design a website, it's possible to thing from the beginning about a "degraded experience" when some third party services are not available, using a "circuit breaker"."

File: news/_posts/2019-21-01-resilient-php-applications.md [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+
+The [Circuit Breaker](https://martinfowler.com/bliki/CircuitBreaker.html) design pattern provides strategies and functions to manage the unavailability of the third party services.
+
+The basic idea behind the circuit breaker is very simple:
+
+* every call of third party service is wrapped into a function call from the Circuit Breaker;
+* once we reach a defined numbers of failures (means the service is unreachable), the Circuit Breaker is "Open": an alternative response is provided and no calls to the service are done;
+* after a defined amount of time (called "reset timeout" sometimes), the Circuit Breaker try to access the service: if the service is available it will be used again, else the alternative response is provided instead;
+
+![https://martinfowler.com/bliki/images/circuitBreaker/sketch.png](https://martinfowler.com/bliki/images/circuitBreaker/sketch.png)
```
Comment (ttoine) [inline news/_posts/2019-21-01-resilient-php-applications.md]: @mickaelandrieu need the link to the original blog article, not only the picture. could you please provide it?

File: news/_posts/2019-21-01-resilient-php-applications.md [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+### How works the PrestaShop Circuit Breaker?
+
+We were not satisfied with the quality and the constraints of the existing libraries or Symfony bundles:
+
+* we want something **really** extensible;
+* we want something **fully tested** and with **high quality** standards;
+* we want something that doesn't require uncommon PHP extensions (like *pcntl*)
+* and ... we still need to provide PHP 5.6+ compatibility;
+
+This is why we have worked on our own implementation of the design pattern, making adapters that relies on the popular HTTP Client [Guzzle](http://docs.guzzlephp.org/en/stable/index.html) to make calls and on the [Symfony Cache](https://symfony.com/doc/3.4/components/cache.html#simple-caching-psr-16) component for the storage/logging of the calls.
```
[the remark this answers is not in the document]
Comment (ttoine) [reply news/_posts/2019-21-01-resilient-php-applications.md]: @mickaelandrieu more precision please

**Review by ttoine — COMMENTED** [review]
needs 2 feedback from @mickaelandrieu

File: news/_posts/2019-21-01-resilient-php-applications.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (mickaelandrieu) [reply news/_posts/2019-21-01-resilient-php-applications.md]: I can't be more precise:

* Guzzle do the HTTP calls,
* Symfony Cache store/log the HTTP calls

How can I reword this to be understandable?

[the remark this answers is not in the document]
Comment (ttoine) [reply news/_posts/2019-21-01-resilient-php-applications.md]: ok for me. @LouiseBonnard ?

**Review by ttoine — APPROVED** [review]

File: news/_posts/2019-21-01-resilient-php-applications.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (ttoine) [reply news/_posts/2019-21-01-resilient-php-applications.md]: thanks :-)

**Comment by ttoine** [comment]
@mickaelandrieu can I let you check that everything is fine after the review? that there is no change in the meaning?
then, we just have to update the header with dates and hour, the filename, and we are good to merge.

**Comment by mickaelandrieu** [comment]
Hello,

sorry totaly forgot to review again this article.

It's good to me, thanks @ttoine :+1:

File: news/_posts/2019-21-01-resilient-php-applications.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+Indeed, we need to create both resilient systems and websites able to recover quickly from network disasters.
+
+## The risks of depending on external services
+
+Assuming that all external services will work just fine is risky because you don't think about what could happen in case of unavailability:
+
+* If the payment APIs are not reachable, instead of freezing your shop, you could suggest your customer to come back once the system is ok and offer them a little discount;
+* You ought to be able to install PrestaShop even if translations APIs are not reachable, at least in English;
+* You ought to be able to access your back office, no matter how many external APIs respond.
```
Comment (PierreRambaud) [inline news/_posts/2019-21-01-resilient-php-applications.md]: `to access to`?

Comment (LouiseBonnard) [reply news/_posts/2019-21-01-resilient-php-applications.md]: Nope, on dit 'to access something'. ;-)
